### ISSUE TITLE
Batch publishing

### ISSUE THREAD

**Issue by prolic** [pdezwart/php-amqp#211]
It would be a nice addition to allow batch publishing like in https://github.com/php-amqplib/php-amqplib/.

see: https://github.com/php-amqplib/php-amqplib/blob/master/PhpAmqpLib/Channel/AMQPChannel.php#L1128
and https://github.com/php-amqplib/php-amqplib/blob/master/PhpAmqpLib/Channel/AMQPChannel.php#L1144

**Comment by zaq178miami** [comment]
Can you post some benchmarks to demonstrate benefits from batch publishing in php-amqplib and also compare it with publish from php-amqp?

**Comment by prolic** [comment]
Yes, it's here: https://gist.github.com/prolic/94d6c8d2ce2c3922f6a8

ext-amqp is incredible fast :-)

**Comment by pinepain** [comment]
**TL;DR**

 batch publishing doesn't make sense in php-amqp internals. If someone need similar feature it can be implemented in PHP userland code. 

**Explanation**

Batch publishing is not AMQP nor RabbitMQ feature but rather a convenient way in php-amqplib to publish bunch of messages in a row. As I see it, packing message from PHP representation into AMQP representation tooks some significant time in php-amqplib, so it makes sense to first encode messages and then flush them into socket. php-amqp uses under the hood [rabbitmq-c](https://github.com/alanxz/rabbitmq-c) (aka librabbitmq), which is C library to communicate with AMQP-compatible broker targeted RabbitMQ. While batch publishing doesn't make much sense from performance point of view, it also not trivial to implement from how rabbitmq-c internals work. If one need to do batch sending, they can implement it on top of existent php-amqp extension in PHP userland, say, like adding messages from business logic loop to some storage, e.g. array, and after that iterate over that storage and publish all that messages one by one. While `basic.publish` is async method that doesn't have a response from server from IO point of view it is just constant writing to socket, and taking in mind that we have [Nagle's algorithm](https://en.wikipedia.org/wiki/Nagle's_algorithm) disabled, it doesn't make sense to do batch sending.

**Comment by prolic** [comment]
Many thanks for your explanation.
